Adding relevant devcontainer features

JPhSachs opened the pull request:
Cherry-picking commits related to core devcontainer functionality.

Closes #8

sonnivs commented:
Looks good overall. Only issue I see is that all of the checks failed on this PR, which seems to be because of a broken `poetry.lock` file, that is not consistent with the `pyproject.toml` state. That probably happened because of the merging/cherry picking.
I see that the next PR fixes that and thus fixes all the checks except for the lint checker.
We could either fix both things now before we merge this or decide to ignore this for now and just fix everything with the next PR, just to get to a good base state to start working on everything else. Up to you 😊

JPhSachs commented:
Review suggestion for this PR:
1) Install Docker + VS Code + "Dev Containers" extension for VS Code.
2) Choose "Rebuild and reopen in container" from the command palette.
3) Wait for the build to finish successfully.

JPhSachs commented:
Changing scope slightly to using prebuilt container images that are pulled from a container image registry (Docker Hub).

Requires modified workflow for:
a) Building the containers automatically from within GitHub and uploading them to Docker Hub
b) Inclusion into the `.devcontainer/devcontainer.json` settings

JPhSachs commented:
1) Added workflow to build the devcontainer images through GitHub Actions, through `.github/workflows/devcontainer.yml`.
2) Replaced the source of the devcontainer to be the corresponding Docker Hub image instead of a local build.

REMARKS:
- Right now, the workflow to build an updated devcontainer image has to be triggered manually. This is to avoid the build of a new image upon every push (merge) to the master branch. In the yaml file, there are already some code snippets looking for a *diff* in the poetry.lock file, but currently commented out. This feature can be added in a future iteration.
- In order to test the workflow feature with manual trigger, I figured out that there must be a yaml file with the exact same name on the master branch. Thus, there are a few commits directly on `master`, but only affecting these files: First, a few to add test cases, then one to remove them, and finally one to add the final file whose name is `.github/workflows/devcontainer.yml`
- I pulled the `poetry.lock` and `pyproject.toml` files from master to reflect their current status when building the devcontainer. However, I also added the `pytest-cov` library for the codecov features.

**When reviewing this PR, please check:**
a) Whether you can correctly pull the devcontainer from Docker Hub with the new settings (instead of trying a local build).
b) Whether the `poetry.lock` and `pyproject.toml` files need further modifications (and if yes, which?) for a clean merge.

JPhSachs commented:
Now with _linux/arm/v8_ and _amd64_ – successfully tested by a developer who had nothing to do with the repository yet.
